nixos/modules/services/network-filesystems/ipfs-cluster.nix

```diff
@@ -0,0 +1,136 @@
+        type = lib.types.str;
+        default = "ipfs";
+        description = "Group under which the ipfs-cluster daemon runs.";
+      };
+
+      consensus = lib.mkOption {
+        type = lib.types.enum [
+          "raft"
+          "crdt"
+        ];
+        description = "Consensus protocol - 'raft' or 'crdt'. https://cluster.ipfs.io/documentation/guides/consensus/";
+      };
+
+      dataDir = lib.mkOption {
+        type = lib.types.str;
+        default = "/var/lib/ipfs-cluster";
+        description = "The data dir for ipfs-cluster.";
+      };
+
+      initPeers = lib.mkOption {
+        type = lib.types.listOf lib.types.str;
+        default = [ ];
+        description = "Peer addresses to initialize with on first run.";
+      };
+
+      openSwarmPort = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = "Open swarm port, secured by the cluster secret. This does not expose the API or proxy. https://cluster.ipfs.io/documentation/guides/security/";
+      };
+
+      secretFile = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = ''
+          File containing the cluster secret in the format of EnvironmentFile as described by
+          <citerefentry><refentrytitle>systemd.exec</refentrytitle>
+          <manvolnum>5</manvolnum></citerefentry>. For example:
+          <programlisting>
+          CLUSTER_SECRET=<replaceable>...</replaceable>
+          </programlisting>
+
+          if null, a new secret will be generated on first run.
+          A secret in the correct format can also be generated by: openssl rand -hex 32
+        '';
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.enable -> config.services.kubo.enable;
+        message = "ipfs-cluster requires ipfs - configure and enable services.kubo";
+      }
+    ];
+
+    environment.systemPackages = [ pkgs.ipfs-cluster ];
+
+    systemd.tmpfiles.rules = [ "d '${cfg.dataDir}' - ${cfg.user} ${cfg.group} - -" ];
+
+    systemd.services.ipfs-cluster-init = {
+      path = [
+        "/run/wrappers"
+        pkgs.ipfs-cluster
+      ];
+      environment.IPFS_CLUSTER_PATH = cfg.dataDir;
+      wantedBy = [ "default.target" ];
+
+      serviceConfig = {
+        ExecStart = [
+          "${pkgs.ipfs-cluster}/bin/ipfs-cluster-service init --consensus ${cfg.consensus} ${initFlags}"
+        ];
+        Type = "oneshot";
+        RemainAfterExit = true;
+        User = cfg.user;
+        Group = cfg.group;
+        EnvironmentFile = lib.mkIf (cfg.secretFile != null) cfg.secretFile;
+      };
+      unitConfig.ConditionDirectoryNotEmpty = "!${cfg.dataDir}";
+    };
+
+    systemd.services.ipfs-cluster = {
+      environment.IPFS_CLUSTER_PATH = cfg.dataDir;
+      wantedBy = [ "multi-user.target" ];
+
+      wants = [ "ipfs-cluster-init.service" ];
+      after = [ "ipfs-cluster-init.service" ];
```

Comment:
I take it this works with the condition on `ipfs-cluster-init`? Since it won't run once the data directory is initialized.

Reply:
Yep it does, I've tested this.